### PR TITLE
ref(injection): Rebuild `fixup_js_file` on top of `magic-string` and sourcemap composition

### PR DESCRIPTION
**Pull request by loewenheim**
This makes debug id injection into JS files precise.

1. Find the point where we need to inject by means of a regex that skips a hashbang, comments, empty lines, and a `"use strict";` statement.
2. Use `magic-string` to insert the debug id code snippet at the injection point and append the `//# debugId` comment.
3. Have `magic-string` give us a sourcemap that maps between the injected and original file. This is really only relevant for the code snippet, the comment is just appended and shouldn't affect mappings anyway.

The sourcemap obtained in (3) is then used to rewrite the original sourcemap using `SourceMap::adjust_mappings`.

This also minorly refactors `fixup_js_file_end`.

### CODE REVIEW

File: src/utils/sourcemaps/inject.rs [hunk 1]
```diff
@@ -763,4 +606,21 @@ more text
...
+  // some other comment
+ "use strict"; rest of the line
+'use strict';
+some line
+//# sourceMappingURL=fake2
+//# sourceMappingURL=real
+something else"#;
+
+        let m = PRE_INJECT_RE.find(source).unwrap();
+        dbg!(&source[m.range()]);
```
Comment (kamilogorek) [inline src/utils/sourcemaps/inject.rs:624]: Should we actually assert on something here?

Comment (loewenheim) [reply src/utils/sourcemaps/inject.rs:624]: Forgot to remove that test when I was writing this. I don't think it serves much of a purpose.

File: src/utils/sourcemaps.rs [hunk 2]
```diff
@@ -790,134 +790,166 @@ impl SourceMapProcessor {
...
-
-                            inject::insert_empty_mapping(&mut decoded)?;
-                            let encoded = data_encoding::BASE64.encode(&decoded);
-                            let new_sourcemap_url = format!("{DATA_PREAMBLE}{encoded}");
+            // Modify the source file and the sourcemap.
+            // There are seveal cases to consider according to whether we have a sourcemap for the source file and
+            // whether it's embedded or external.
+            let debug_id = match sourcemap_url {
+                None => {
+                    // Case 1: We have no sourcemap for the source file. Hash the file contetns for the debug id.
```
Comment (kamilogorek) [inline src/utils/sourcemaps.rs:798]:
```suggestion
                    // Case 1: We have no sourcemap for the source file. Hash the file contents for the debug id.
```

File: src/utils/sourcemaps.rs [hunk 3]
```diff
@@ -790,134 +790,166 @@ impl SourceMapProcessor {
...
-                            // Update the embedded sourcemap and write it back to the source file
-                            let Ok(mut decoded) = data_encoding::BASE64.decode(encoded.as_bytes()) else {
-                                bail!("Invalid embedded sourcemap in source file {source_url}");
-                            };
-
-                            inject::insert_empty_mapping(&mut decoded)?;
-                            let encoded = data_encoding::BASE64.encode(&decoded);
-                            let new_sourcemap_url = format!("{DATA_PREAMBLE}{encoded}");
+            // Modify the source file and the sourcemap.
+            // There are seveal cases to consider according to whether we have a sourcemap for the source file and
```
Comment (kamilogorek) [inline src/utils/sourcemaps.rs:794]:
```suggestion
            // There are several cases to consider according to whether we have a sourcemap for the source file and
```

**Review by kamilogorek — APPROVED** [review]
That's a beefy change, fingers crossed🤞

**Pull request merged**